src/api/regex.c

```diff
@@ -84,18 +208,121 @@ static int f_pcre_match(lua_State *L) {
     we just detect this case and give up. */
     luaL_error(L, "regex matching error: \\K was used in an assertion to "
     " set the match start after its end");
+    if (regex_compiled) pcre2_code_free(re);
     pcre2_match_data_free(md);
     return 0;
   }
   for (int i = 0; i < rc*2; i++)
     lua_pushinteger(L, ovector[i]+offset+1);
+  if (regex_compiled) pcre2_code_free(re);
   pcre2_match_data_free(md);
   return rc*2;
 }
 
+static int f_pcre_gmatch(lua_State *L) {
+  /* pattern param */
+  bool regex_compiled = false;
+  pcre2_code* re = regex_get_pattern(L, &regex_compiled);
+  if (!re) return 0;
+  size_t subject_len = 0;
+
+  /* subject param */
+  const char* subject = luaL_checklstring(L, 2, &subject_len);
+
+  /* offset param */
+  size_t offset = regex_offset_relative(
+    luaL_optnumber(L, 3, 1), subject_len
+  ) - 1;
```

Comment:
I wonder if we should consider the offset as UTF-8 char offset. Probably not, as we're already returning byte offsets from other regex functions.

Reply:
Not sure either, depends on the use case I guess... but if that would be the case we would also need to check for bom and skip it, I would say leave it as is and maybe add some regex.ugsub if needed at a later time.